### ISSUE TITLE
Proposal: Change to the Management of Issues and Pull Requests

### ISSUE THREAD

**Issue by nrmay** [mytardis/mytardis#873]
# Background
First, since some of developers on the MyTardis project may not have an academic software engineering background, I shall provide some context. The Software Product Lifecycle (SPLC) [1, p8-4] is the sequence of steps that make up the software engineering process. The steps of the basic, simplified process are as follows:

   Requirements → Design → Construction → Testing → Deployment → Support.

Software engineering methodologies (including those that can be characterized as Agile, Spiral, Waterfall) all implement these steps, in some form. Best practice dictates that the steps must be performed in sequence from left to right if a successful outcome is to be achieved, however, jumping backwards is often allowed for iteration or fixing issues discovered in a later step.

MyTardis is a multi-institution, open-source project, consisting of multiple applications and their repositories. Historically, the repositories have been managed using the 'Benevolent Dictator' model [2, p72] where administration privileges were restricted to a small set of active committers at a single institution. The diverse components have been developed largely independently and sometimes with duplication of effort, as there has been little co-ordination of development efforts across institutions.
# Problems

* Important design decisions often delayed or unresolved due to disagreements over best practice. This stems from several factors:  
  * A broader discussion of design often does not happen until after a pull request has been submitted.
  * A neutral, balanced discussion cannot be had when one party has already invested significant effort into the implementation of their preferred design solution. In other words, _Collaborative Design_ is happening after _Construction_.
* No coherent high-level design.
  * i.e. no architecture analysis and design.
* Pull Requests often not merged promptly.
* Pull Requests closed by third parties and re-implemented prior to merging.
* Interdependency of concurrent changes not identified proactively.
  * Leading to significant remediation required when important pull requests are merged.

# Proposed Solution

1. Fixed Release Cycle:
   - period to be determined, so as to be responsive to stakeholder requirements.

2. Proposals must be raised as Issues (not PRs).

3. Proposals may only be accepted by a vote:  
   -  method to be determined, so that a consensus can be achieved across institutions.

4. Pull Requests can only address:  
   - Bug Fixes, 
   - Dependency Updates, or 
   - Accepted Proposals from prior releases.

5. Pull Requests to be reviewed before merging:  
   -  method to be determined, so as to promote knowledge transfer and coherent architecture/design.

6. Merge privileges to be shared across all institutions.

# Benefits

- More responsive release cycles.
- Greater visibility of change schedule.
- Better management of migration impact and planning.
- Faster turnaround on important design decisions.
- Broader ownership of the codebase.
- Acknowledgement of contributions.
- Improved knowledge transfer.
- Better architecture.
- Will support the creation of a MyTardis roadmap.

# Risks

- The proposed solution only considers the management of the Github repository, and not the broader MyTardis project.
- The voting method chosen must be sustainable, so as to avoid backup in the approval process.

# References
[1] Guide to the Software Engineering Body of Knowledge (SWEBOK), Version 3.0, IEEE Computer Society, 2014, [SWEBOKv3.pdf](http://www4.ncsu.edu/~tjmenzie/cs510/pdf/SWEBOKv3.pdf) 

[2] Fogen, Karl, Producing Open Source Software : How to Run a Successful Free Software Project, 2005-2016, CC-BY-SA (4.0), [Producing Open Source Software](http://producingoss.com/en/producingoss.pdf)

**Comment by wettenhj** [comment]
@nrmay, thanks for your passion and desire to make MyTardis better!  It would be awesome if we could work together more effectively to add value to MyTardis which will benefit all of us!  I strongly agree with some of your suggestions, but strongly disagree with other suggestions, so I couldn't vote for this proposal in its current form.

One book I would recommend on this topic is [Managing Open Source Projects, A Wiley Tech Brief](http://au.wiley.com/WileyCDA/WileyTitle/productCd-0471189170.html) (available on Google Play).

### Fixed Release Cycle

The suggestion I agree with most strongly is the idea that a Fixed Release Cycle would be useful for MyTardis, particularly for users outside of Monash.  Inside Monash, we prefer to deploy tags from the develop branch, because being up-to-date is more important than being stable for our stakeholders, and if something goes wrong, one of us will respond to a Help Desk request or Nagios alert or Django email alert and we'll post the issue on Slack, then submit a PR to fix it, and have another MyTardis core developer review and merge the PR.

However for users outside of Monash, we can't guarantee that recent changes in the develop branch are 100% compatible with recent changes in external forks of MyTardis and related apps at deployment sites outside of Monash.

So let's brainstorm how we could create stable MyTardis releases - here are some suggestions:
1. Assign the role of "release manager" to one of the MyTardis core developers.  Nominees could be proposed at MyTardis Community Meetings and voted on.
2. Set up a schedule, e.g.
- **Feb 1st and Aug 1st:** A calendar alert reminds all MyTardis developers that in one month's time (Mar 1st and Sep 1st), the develop branch will be merged into master in preparation for tagging a stable release.
- **Mar 1st and Sep 1st:** develop is merged into master and a calendar alert reminds all MyTardis developers that a stable release will be tagged from the master branch in one month's time (Apr 1st and Oct 1st) and that only bug fixes and documentation fixes can be merged into master in the lead up to the tagged release.
- **Apr 1st and Oct 1st:** The master branch is tagged for the stable release.  The master branch will usually not be updated again for another 5 months, unless merging a critical bug fix is proposed and agreed to at a MyTardis Community Meeting.

### Guidelines for Acceptable Pull Requests

The thing I disagree with most strongly is the idea that Pull Requests into develop must follow strict rules, i.e. "Proposals must be raised as Issues (not PRs)" and "Pull Requests can only address: ...".  I have no problem having stricter guidelines about what types of Pull Requests are appropriate for the master branch, but I don't think there should be such strict guidelines for PRs against the develop branch.

Discussions after a pull request is submitted are a natural part of the GitHub Flow: https://guides.github.com/introduction/flow/

Some developers may feel that the PR review discussions include design decisions which they would have liked to have been consulted on earlier.  But we can't change history, so generally PR discussions should be constructive and suggest ways to move forward, which might require suggesting compromises.

The third page of the "GitHub Flow" guide ("Open a Pull Request") says:

> "You can open a Pull Request at any point during the development process: when you have little
> or no code but want to share some screenshots or general ideas, when you're stuck and need
> help or advice, or when you're ready for someone to review your work".

**Comment by nrmay** [comment]
@wettenhj 
I'm having some difficulty understanding your development methodology.
> Inside Monash, we prefer to deploy tags from the develop branch, because being up-to-date is more important than being stable for our stakeholders,...

- Are you saying that Monash runs all its production MyTardis deployments from the develop branch of the core mytardis repository?
- What is your average turnaround from change request to production deployment?
- Do you verify changes on User Acceptance Test deployments? (we call them Non-Production Environments: NPEs)

> have another MyTardis core developer review and merge the PR.

- Who are the 'core developers'?

> However for users outside of Monash, we can't guarantee that recent changes in the develop branch are 100% compatible with recent changes in external forks of MyTardis and related apps at deployment sites outside of Monash.

- It's worse that 'can't guarantee'. It often generates more work for us, because of the lack of regard for the consequences to other deployments. Adding a fixed release cycle doesn't help if there is no consultation on the changes that are merged into develop: it just tells us when the additional workload is coming.

> Some developers may feel that the PR review discussions include design decisions which they would have liked to have been consulted on earlier.

- Who owns the code?

**Comment by grischa** [comment]
Thank you for your proposal, @nrmay. Your post raises some valid points and it turns out I have been working on something that will solve most of these issues.

The main reason MyTardis is in the state it is currently in can be found in its history of being worked on through project funding from many different sources. Different project teams with different management had varying priorities and often were under pressure to get things done before funding ends. At the height of development, no one was there to push back and set a direction, no one was asked to do so.

In the past, I also was complacent in providing much guidance for the future, mainly because of the typical short-term project focus. I.e. we cannot promise anything in the long-term given our funding requires quick, targeted results and we have no bandwidth to deliver more.
I deliberated and put my thoughts into a roadmap document that outlines the direction MyTardis development is going to go. I will post it on Slack later today. It is a proposal and I want to engage the community around it, hoping to make it a plan everyone can get behind.

Regarding the solutions you proposed, I’m not sure they would actually address the issues very well. However, I will add them to the meeting agenda and I am happy to discuss the details there.

**Comment by nrmay** [comment]
@grischa I intend only to answer questions on the proposal at the community meeting. I'd be grateful if you could discuss you reasons here as to why you think this proposal won't lead to the benefits outlined.

**Comment by wettenhj** [comment]
@nrmay, I don't have time to reply adequately to all of your questions and concerns, but I've managed to come up with a short essay in response to this one 😄 :

> It's worse that 'can't guarantee'. It often generates more work for us, because of the lack of regard for the consequences to other deployments. Adding a fixed release cycle doesn't help if there is no consultation on the changes that are merged into develop: it just tells us when the additional workload is coming.

I think the "lack of regard" and "no consultation" sentiments are unfair - we use an open-source model where all changes are transparent and you are free to discuss risks you see in code changes before you deploy them.

Regarding your concern that "a fixed release cycle doesn't help" unless you can veto changes to the develop branch, I disagree.  If you're concerned that the next fixed release will break your current deployment, you can try an alpha or beta version shortly before the scheduled release date and provide feedback - then we can determine whether the problem with your attempted upgrade represents a serious flaw in the pending release and if so, either fix it in the pending release, or propose a minor / patch release shortly afterwards.

Perhaps you could elaborate on your concern that changes in MyTardis often generate more work for you?  Specific examples could be useful.  I don't doubt that you have been burned by breaking changes at times, but it's worth understanding the scale of the problem.  For example, if fixing your deployment in response to an unexpected change in MyTardis took one hour, would it have taken more than an hour to have had a discussion with you about the proposed change (and any other changes which could have potentially led to problems with your upgrade) beforehand?  Would MyTardis development slow down significantly if every change was discussed in detail with everyone who wants to be consulted?  Given that most recent changes have been from Monash contributors, ensuring everyone is consulted adequately about every change would create more work for Monash developers - is there anything being offered in return?

If you can come up with a specific example or two of the "more work for us" problems you refer to,
- How long did it take you to resolve the problem?
- Did you ask anyone for help?  If so, were you satisfied with their response?
- How often was that MyTardis deployment upgraded? (incremental upgrades are easier).
- What would have been your preferred way to have been consulted about the change beforehand?

Fortunately if you come across a problem with upgrading MyTardis, you can usually get a quick response on the MyTardis Slack.  And you can reduce the impact of the problem by upgrading more often.  You can also help by offering to review pull requests in a constructive manner or by submitting more pull requests yourself (which need to be rebased against develop, which can help you to keep up-to-date with understanding the latest code changes in develop).

But sometimes we won't all agree on the risk versus reward of a PR.  Having some disagreements is a normal part of open source development, but none of us like disagreements leading to stalled Pull Requests like the "BDD Tests" PR (which is still not merged and no alternative PR has been proposed).  So I would prefer to have these stalled PRs merged or closed, either by asking @grischa to make the final decision as the project's de facto "benevolent dictator" or by creating a dispute resolution committee (perhaps made up of one member from each university which has made significant contributions to MyTardis in recent years) which can vote to resolve the disagreements.  At Monash, we require at least two MyTardis developers / sys admins to be involved in any change affecting our deployments or MyTardis's "develop" branch (one submits the PR and the other merges it).  In an ideal word, every MyTardis user and developer would endorse every PR before it is merged, but that just isn't practical - we just don't have the resources to have multiple people giving detailed reviews of every PR, and requiring merge approval from more collaborators would create even more stalled PRs which I don't think anybody wants to see.

As @grischa has mentioned in his comment above, there is a lot of technical debt in MyTardis - code which is in desperate need of tidying up, refactoring, and/or having tests written for it.  I think we all agree that having better code quality and improved test coverage would make MyTardis easier to maintain in the long term.  But some of the refactoring required to improve the code quality would be more than just "minimal bug-fix changes", so if you have an extremely low tolerance for changes which risk introducing new bugs (e.g. from refactoring), then you might find it difficult to collaborate on a single MyTardis branch ("develop") with other MyTardis developers who have an interest in reducing technical debt.  But in the spirit of collaboration, it would be preferable for you to make contributions to the core develop branch and accept that on some occasions, someone will unintentionally introduce a bug or incompatibility which will affect you.

**Comment by nrmay** [comment]
@wettenhj Thanks for the essay, but answering the questions would have been more useful to me.

**Comment by wettenhj** [comment]
@nrmay,

OK, answers to questions below:

> Are you saying that Monash runs all its production MyTardis deployments from the develop branch of the core mytardis repository?

Yes - each time an upgrade is needed, the HEAD of develop is considered the default candidate for the upgrade, and if it looks suitable, it is tagged and deployed, first in a Non-Production Environment and then if it passes the required tests, it is deployed in production.

> What is your average turnaround from change request to production deployment?

A few weeks would be typical, but of course it varies hugely depending on the impact, urgency and complexity of the change request.

> Do you verify changes on User Acceptance Test deployments? (we call them Non-Production Environments: NPEs)

Yes of course.

> Who are the 'core developers'?

Sorry for the vague term - I just meant anyone that the repository owner (@grischa) trusts to review and merge pull requests.

> Who owns the code?

I'm not a lawyer, so I can't give you a legal answer.  Grischa is the repository owner in the GitHub sense.  I imagine that most MyTardis contributors have signed intellectual property agreements with their employers which give their employers some degree of ownership.

**Comment by grischa** [comment]
@nrmay, your issues are symptoms of two limitations of the MyTardis project:
- lack of resources
- technical debt

Adding all these requirements would strain the available resources further. They are not addressing technical debt, they would only slow down positive changes.
Imagine a project funded to add a feature. The feature gets proposed, it is well described in coarse detail in the proposal and gets approved. It gets developed by one person, then at the end of the project there is a pull request. The code is terrible and spaghetti, but it gets merged regardless because it’s the only project that puts any money into MyTardis at the time. I don’t see where your process helps here, except create more work.

Regarding your specific suggestions:

>Fixed Release Cycle:

I would agree to a monthly release cycle.
As James mentioned, we work in a continuous release mode at Monash. The reason is mainly that there haven’t been any outside contributions recently that needed lengthy testing or release decisions to be made. I think it would be best if we formalise this by periodically tagging the state of develop, e.g. as a monthly release.
I know you don’t like this. The roadmap I will post lays out a path to a different state of affairs, with fewer releases and independent deployments.
However, given our scarce resources, it is likely that we are not able to keep supporting a bunch of release branches with bug fixes and dependency updates. The solution isn’t more process, but moving forward. By strengthening the deployment pipeline, it will become easier to update than to maintain an old installation. The world is moving this way, see cloud software offerings and subscriptions models for things like Adobe or JetBrains.

>Proposals must be raised as Issues (not PRs).

Disagree. Sometimes writing a bit of code is easier than describing it, sometimes some code needs to be written for proof-of-concepts, internal tests or funded projects and a rejected or amended proposal would not affect whether and how the work gets done. If we were using SVN or something hard to manage, I might agree, but for git and GitHub throwing around code and pull requests is really their greatest aspect.
Also, a proposal acceptance doesn’t give anyone a free pass to get their PR merged. There shouldn’t be an assumption that code can’t be rejected when the overall idea is sound.

>Proposals may only be accepted by a vote:

While not formalised, existing proposals/PRs have so far not been accepted without proper discussion when there was any kind of concern or disagreement. Waiting for feedback has been holding back PRs, e.g. the BDD PR, and the same will happen with voting, because a vote is basically feedback. I would agree to a voting system if it comes with time limits. E.g. if there hasn’t been any vote/feedback within 10 days, PR gets closed, if there have been votes and no further activity within 10 days, the latest vote result will stand. We can discuss the details.

>Pull Requests can only address:

See my answer to “Proposals must be raised as Issues (not PRs)”

>Pull Requests to be reviewed before merging:

Yeah, I think that requirements for contributions can be improved. We can collect some criteria for proper commenting of code, documentation and tests.
For knowledge transfer I find Slack to be very useful. You will not find any volunteer to document or comment on all the hidden assumptions in the current code-base. Looking forward though, let’s try to do better for new code contributions.

>Merge privileges to be shared across all institutions.

Happy to give people access to mytardis/mytardis repo branches. In fact, I have been encouraging my colleagues to develop on branches in the main repo for better visibility. You can be your own branch maintainer.
However, develop/master branch merging will remain in the hands of the current admins as it reflects experience and contributions.
We might consider changing this in the future, of course. The project belongs to the people who contribute constructively.

Regarding the benefits:

> More responsive release cycles.

Seems like we are on the same page here, frequent releases are good.

> Greater visibility of change schedule.

Where do you see greater visibility coming from this? In my experience, posting proposals has not led to improved visibility.

> Better management of migration impact and planning.

A lack of resources might not allow us to manually consider everyone’s specific situation. If you could answer James’ questions about issues you had in the past, maybe we will better understand your concerns.
The best way to avoid regressions is automatic testing.

> Faster turnaround on important design decisions.

Your solutions don’t address fast turnaround, they may even slow it down.

> Broader ownership of the codebase.

Is this a benefit or a problem? The past has shown that broad ownership leads to spaghetti. Broad contributions and confident leadership appear to be a better model.

> Acknowledgement of contributions.

Git and GitHub automatically acknowledge contributions

> Improved knowledge transfer.

Open source is open, and Slack is available. Most questions get answered within hours if not minutes there, and we have almost-monthly catch ups where we can chat about anything. Sure, a nice proposal can help, I’m not denying that, but asking for clarification is a close to zero extra cost.

> Better architecture.

You are not proposing any architecture improvements

> Will support the creation of a MyTardis roadmap.

Incidentally, one has already been created. I got caught up with some other work today, so I’ll post it tomorrow.

Finally, for further and more constructive discussions please create separate issues for each of your mostly independent proposals. This thread is becoming lengthy and hard to follow as each post addresses a number of concerns.

**Comment by nrmay** [comment]
@grischa @wettenhj Thank you both for your comments. They have been most enlightening. However, it's clear to me that we are not in the same book, let alone on the same page. Therefore, I'll close this issue (mischief done!).
